decentraland/unity-renderer
# feat: ab detection functionality

## What does this PR change?

This PR:

1. Adds a tool for highlighting which mesh renderers are rendering the content loaded with asset bundles and which ones are loaded with raw GLTFs. DebugBridge receives a call from Kernel triggered by the chat command. This calls is passed to a ABDetectorPlugin via DataStore values as an intermediary. ABDetectorPlugin loads dedicated debug materials lazily on first call and applies them to the whole scene.
2. Adds custom low level data structure Multimap which is a generic wrapper for a dictionary of lists. After discussion with Alex we figured there are no class like that and decided to try using this one since it allows to not repeat a lot of code.

[This PR](https://github.com/decentraland/unity-renderer/pull/2058) was used as a reference for implementation.

Closes #2911

This Kernel PR implements chat commands for this feature:
https://github.com/decentraland/kernel/pull/536

New chat commands are:

/detectabs => enable for all shapes
/detectabs off => disable for all shapes
/detectabs scene => enable for current scene shapes
/detectabs scene off => disable for current scene shapes (works only if "/decetabs scene" was used for this scene to enable)

...

## How to test the changes?

This PR has to be tested together with the corresponding Kernel PR. To do this, use these links:

https://play.decentraland.zone/?renderer-branch=ab-detection-functionality&kernel-branch=ab-detection-functionality

1. Use chat command `/detectabs ` at some place with ABs enabled. The shapes should turn green.
2. Use chat command `/detectabs off` . The shapes should turn to their usual look.
3. Use chat command `/detectabs scene` . The shapes should turn green only within current scene.
4. Use chat command `/detectabs scene off` . The shapes of that last scene should turn to their usual look.
5. Command `/detectabs off` should also turn all shapes to normal look after  `/detectabs scene`

6. Repeat 1-2 with ABs disabled, the color should be red instead of green. To test with asset bundles disabled use this link:
https://play.decentraland.zone/?renderer-branch=ab-detection-functionality&kernel-branch=ab-detection-functionality&DISABLE_ASSET_BUNDLES=

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
